### PR TITLE
WIP: Initial version of base Connection classes

### PR DESCRIPTION
**Pull request by datapythonista**
xref #2678 

Starting the definition of the backend base classes. Planning to do it iteratively, otherwise this will end up being a huge and complex PR, while the changes should be simple.

The idea of these new classes is to be the main base class for backends, replacing `Backend` and `Client`. All the methods defined in them already exist in those two classes, just reorganizing in a better way.

I think we should do some clean up of the API before the release. But for now just keeping what's in Ibis, and when the new base classes are finished and used by the backends we can discuss. Examples of things I'd change:
- Merge `execute` and `raw_sql`, and have a single method to execute, which accepts an expression or a string (or an AST if useful). Same for the schema methods, which we have one for schema from expression, and schema from AST.
- Simplify API to interact with databases (see #2804)
- Make some methods private. Some are public because they are public for backend authors who need to implement them, but shouldn't be exposed to final users
- Delete duplicate stuff, and leave a single way to execute things (see #2793)
- Just an idea, but maybe we could group DML methods under a `DML` class or something, instead in the top level, which is already a huge namespace

The new classes aren't used yet, so there may be errors, that will be fixed in follow ups, since this is not yet being tested.

We also need to consider that we are not running mypy (see #2823). I'm adding types here, which should be mostly correct, but nothing is being validated. And enforcing types is a decent effort, not sure if something to do before or after the release

@jreback let me know if any thoughts about the above

### CODE REVIEW

**Comment by jreback** [comment]
why are we changing the name from Backend? that is a pretty informative name.

**Comment by datapythonista** [comment]
> why are we changing the name from Backend? that is a pretty informative name.

I'm fine with `Backend` too, I explain why I think `Connection` it's better in the issue, #2678, at the end of the description.

File: ibis/backends/base/connection.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            Spark based backends.
+        """
+        pass
+
+    @abc.abstractmethod
+    def connect(connection_string: str, **options) -> BaseConnection:
+        """
+        Connect to the underlying database and return a client object.
+        """
+        pass
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:49]: I don't think we need `pass` for abstractmethod?

File: ibis/backends/base/connection.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    def register_options(self) -> None:
+        """
+        If the backend has custom options, register them here.
+        They will be prefixed with the name of the backend.
+        """
+        pass
+
+    def compile(self, expr: ir.Expr, params=None) -> str:
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:58]: I think only SQL backend would return a str for this. So many should change this to `Any` in base class.

File: ibis/backends/base/connection.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        """
+        Compile the expression.
+        """
+        return self.compiler.to_sql(expr, params=params)
+
+    def verify(self, expr: ir.Expr, params=None):
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:64]:
```
def verify(self, expr: ir.Expr, params=None) -> bool:
```
?

File: ibis/backends/base/connection.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        in SQL backends, a NullLiteral operation could be translated simply
+        with the string "NULL".
+
+        Examples
+        --------
+        >>> @ibis.sqlite.add_operation(ibis.expr.operations.NullLiteral)
+        ... def _null_literal(translator, expression):
+        ...     return 'NULL'
+        """
+        if not hasattr(self.client, 'compiler'):
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:92]: Maybe better to leave an empty implementation here and implement in SQL backends?

File: ibis/backends/base/connection.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                operation, translation_function
+            )
+
+        return decorator
+
+    @property
+    def current_database(self) -> str:
+        """Return the current database."""
+        # TODO: If we can assume this, we should define the `con` attribute
+        return self.con.database
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:108]: What is `self.con`?

File: ibis/backends/base/connection.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    def database(self, name: Optional[str] = None):
```
Comment (icexelloss) [inline ibis/backends/base/connection.py:110]: type annotate return type?

File: ibis/backends/base/sql/connection.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        Returns
+        -------
+        Backend cursor
+        """
+        cursor = self.con.execute(query)
+        if cursor:
+            return cursor
+        cursor.release()
+
+    def execute(self, expr, params=None, limit='default', **kwargs):
```
Comment (icexelloss) [inline ibis/backends/base/sql/connection.py:43]: Perhaps this method can be lifted to base class?

Comment (icexelloss) [reply ibis/backends/base/sql/connection.py:43]: (Since all backend supports execution)

**Comment by icexelloss** [comment]
Thanks @datapythonista !

I like the idea of union "Backend" and "Client" since I think they are pretty much the same thing. But I am with @jreback that I think `Backend` is better. `Connection` to be seems to be a pretty simple/"dumb" class just from how the word is used in, for example, "HTTP Connection". I feel `Backend` is much more complex as it could do pretty confiscated work like translate to SQL or to a PySpark DataFrame. It's kind of a "complier".

Orthogonally, since this looks like breaking change if you are going to remove the existing Client and Backend classes in favor of this one. So I am curious about your thinking around when is a stable time to refactor implementation such as ibis-bigquery to work with newest ibis. Internally we can no longer pull the ibis default (we cherry pick from ibis default now to keep big-query working) because we need both ibis and ibis-bigquery at the moment.

**Review by icexelloss — CHANGES_REQUESTED** [review]
Some comments

**Comment by datapythonista** [comment]
Thanks for all the comments @icexelloss, very useful. Will address comments and reply more in detail later. But regarding backward compatibility: This doesn't break compatibility with users (so far), or the previous changes. The interface with third-party backends has been broken for a while with the refactoring, but I'm working on porting the OmniSciDB backend to the latest Ibis at the moment, and we can quickly do the same for bigquery after that. Sorry it's been annoying, but I don't think it should be a big concern, since it's few hours of work and fully under our control to keep the master of the backends we control in sync. Let me know if I'm wrong or missing something, but if you're using Ibis master in the company (I wasn't aware), we can plan better those things. It'd also be useful for me if you can tell me which backends are you using. I guess PySpark and BigQuery (not sure if Impala), but if it's few we can prioritize those.

Regarding compile/execute, I want to have a look, I don't think pandas should use them more than for compatibility. Not sure if requiring them in the base class is the best we can do.

**Comment by jreback** [comment]
> > why are we changing the name from Backend? that is a pretty informative name.
> 
> I'm fine with `Backend` too, I explain why I think `Connection` it's better in the issue, #2678, at the end of the description. tl;dr: this class is being returned by `ibis.backend.connect()`, so `Connection` is more intuitive than `Backend` IMHO.

i c, so if we had to write a description (and we should)
- backend is the package of where the code lives
- `Connection` is kind of like Client/Backend now, e.g. its a connection to a database (or equiv for non sql ones).
?

**Comment by datapythonista** [comment]
> i c, so if we had to write a description (and we should)
> 
> * backend is the package of where the code lives
> * `Connection` is kind of like Client/Backend now, e.g. its a connection to a database (or equiv for non sql ones).
>   ?

That's what I would do, yes. I think for end users that terminology should be more intuitive. If their code is something like:
```python
db = ibis.sqlite.connect('geography.db')
con = ibis.sqlite.connect('geography.db')
connection = ibis.sqlite.connect('geography.db')
```
I think the class being returned being a `Connection` makes more sense. Also, that's consistent with [SQLAlchemy Connection](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Connection) and [DBAPI2 Connection[(https://www.python.org/dev/peps/pep-0249/#connection-objects) which are equivalent classes in my opinion.

And agree with the docs, I'll review all backend documentation after the release, add a page on how to implement backends, and also see if in the tutorial or elsewhere it makes sense to add more info on how the backend system work, how to install backends...

File: ibis/backends/base/connection.py (hunk 3 above)
Comment (datapythonista) [reply ibis/backends/base/connection.py:64]: This already exists in Ibis. I'm +1 on removing it, not much value added, but in this PR I'm aiming at not making breaking changes, and deprecate whatever is needed (lots of things IMHO) in dedicated follow ups.

File: ibis/backends/base/connection.py (hunk 4 above)
Comment (datapythonista) [reply ibis/backends/base/connection.py:92]: Good catch. I moved it to the SQL Connection class, I don't think we need an empty method here, I think this is specific of SQL backends.

File: ibis/backends/base/connection.py (hunk 5 above)
Comment (datapythonista) [reply ibis/backends/base/connection.py:108]: It's assumed to be the actual connection with the database, a DBAPI2 `Connection` instance. This is hacky, I added the TODO to take care of this later, but this is how it is right now in master, just copied the method.

File: ibis/backends/base/connection.py (hunk 3 above)
Comment (icexelloss) [reply ibis/backends/base/connection.py:61]: kk

File: ibis/backends/base/connection.py (hunk 5 above)
Comment (icexelloss) [reply ibis/backends/base/connection.py:84]: kk

**Review by icexelloss — APPROVED** [review]
+1

**Comment by datapythonista** [comment]
@icexelloss Thanks for the reviews. There are still pending your comments on whether `compile` and `execute` should be in the base connection or only for SQL backends. I thinkI agree for `execute`, since we can have lazy non-SQL backends like Dask. But I want to have a look at the current behavior. Feels like having `compile` in pandas is more a hack for consistency of methods than a good decision, but I need to research before having an informed opinion.

**Comment by icexelloss** [comment]
Looks good to me to merge.

* A few more thoughts on `Connection` vs `Backend`
1.  `Connection` does make more sense if this is connecting to an actual database. It might make less sense if the backend is `pandas` `spark` or `dask` since those backends doesn't really have a concept of a "connection". (In pandas backend, there isn't connection at all. In Spark, a similar concept is called `SparkSession`). So admittedly, there isn't really an universally good name for the class. And that's why I personally still feel `Backend` is perhaps a better name since it's more generic than `Connection` .

2.  It might feel a bit more natural to say "execute expression with a backend" than "execute expression with a connection". 
But I am OK with calling it `Connection` 

(internally we have a thin wrapped class around this object that we customize slightly and this object is not exposed directly to the users, therefore not a big issue for us what is this called)

* On our internal sync with ibis default (To provide some context)
This is a little annoying but not too bad for us now XD.

We try to pull ibis default changes into our internal environment instead of waiting for ibis releases, since our internal project is in very active development and relies pretty heavily on ibis changes. So currently we are in a state of not able to pull default changes because the compat issue with ibis-bigquery default. This is not a huge issue yet because we have been cherry-picking changes from ibis default and skip the refactor related changes, but I imagine this could break when we start to touch the same piece of code.

> but I don't think it should be a big concern, since it's few hours of work and fully under our control to keep the master of the backends we control in sync

It is currently a bit hard to update big query right now since the default ibis is not released so we need to wait on a new ibis in order to fix bigquery. But agree it's not a big concern.

**Comment by icexelloss** [comment]
> There are still pending your comments on whether compile and execute should be in the base connection or only for SQL backends.

Yeah not a big concern I think and can be addressed follow ups. I agree the `compile` for pandas backend does nothing and can probably raise error (use should never need to call this with pandas backend). Anyway, my thoughts are

`compile` : return a backend specific object, for sql backend, this is `str`. For PySpark backend, this is `pyspark.sql.DataFrame` and for Dask backend, this is `dd.DataFrame`

`execute`: execute the expression and return as a `pd.DataFrame` (bring to memory)

**Review by jreback — CHANGES_REQUESTED** [review]
this should just be called Backend. I think naming this as Connection is just causing more churn and confusion.

**Comment by datapythonista** [comment]
Superseded by #2883

**Pull request closed without merging**